trieste/acquisition/function/function.py

```diff
@@ -1034,6 +1035,608 @@ def __call__(self, x: TensorType) -> TensorType:
+            Sigma_iq = covariance[:, :, q : q + 1]
+            Sigma_qj = covariance[:, q : q + 1, :]
+            Sigma_qq = covariance[:, q : q + 1, q : q + 1]
+
+            cov = Sigma_ij * diq * dqj - Sigma_iq * diq - Sigma_qj * dqj + Sigma_qq
+
+            return cov
+
+        Sigma = tf.map_fn(
+            compute_single_slice,
+            tf.range(Q),
+            fn_output_signature=dtype,
+        )
+
+        Sigma = tf.transpose(Sigma, perm=[1, 0, 2, 3])
+
+        return Sigma
+
+    def compute_p(
+        self,
+        m_reshaped: tf.Tensor,
+        b_reshaped: tf.Tensor,
+        Sigma_reshaped: tf.Tensor,
+        mvn_cdf: Callable[[TensorType, TensorType, TensorType, float], TensorType],
+    ) -> TensorType:
+        """Helper function for the batch expected improvement, which computes
+        the tensor p, as detailed in Chevallier and Ginsbourger
+        :cite:`chevalier2013fast`.
+
+        :param m_reshaped: Tensor of shape (BQ, Q)
+        :param b_reshaped: Tensor of shape (BQ, Q)
+        :param Sigma_reshaped: Tensor of shape (BQ, Q, Q)
+        :returns p: Tensor of shape (B, Q)
+        """
+
+        # Check shapes of covariance tensor
+        tf.debugging.assert_shapes(
+            [
+                (m_reshaped, ("BQ", "Q")),
+                (b_reshaped, ("BQ", "Q")),
+                (Sigma_reshaped, ("BQ", "Q", "Q")),
+            ]
+        )
+
+        # Unpack dtype and mean shape
+        dtype = m_reshaped.dtype
+        BQ, Q = m_reshaped.shape  # (B*Q, Q)
+
+        if BQ % Q == 0:
+            B = BQ // Q
+
+        else:
+            raise ValueError(
+                f"Expected size of dimension 0 of m_reshaped tensor to be "
+                f"divisible by size of dimension 1, instead found "
+                f"{m_reshaped.shape[0]} and {m_reshaped.shape[1]}."
+            )
+
+        # Compute mean, covariance and x for p mvn normal cdf
+        p_cdf_mean = tf.zeros(shape=(BQ, Q), dtype=dtype)  # (B*Q, Q)
+        p_cdf_cov = Sigma_reshaped  # (B*Q, Q, Q)
+
+        p_cdf_x = b_reshaped - m_reshaped  # (B*Q, Q)
+
+        p = mvn_cdf(  # type: ignore
+            x=p_cdf_x,
+            mean=p_cdf_mean,
+            cov=p_cdf_cov,
+        )  # (B*Q,)
+
+        p = tf.reshape(p, shape=(B, Q))  # (B, Q)
+
+        return p
+
+    def compute_c(
+        self,
+        m_reshaped: tf.Tensor,
+        b_reshaped: tf.Tensor,
+        Sigma_reshaped: tf.Tensor,
+    ) -> TensorType:
+        """Helper function for the batch expected improvement, which computes
+        the tensor c, which is the c^{(i)} tensor detailed in Chevallier and
+        Ginsbourger :cite:`chevalier2013fast`.
+
+        :param m_reshaped: Tensor of shape (BQ, Q)
+        :param b_reshaped: Tensor of shape (BQ, Q)
+        :param Sigma_reshaped: Tensor of shape (BQ, Q, Q)
+        :returns c: Tensor of shape (B, Q, Q-1)
+        """
+
+        # Check shapes of covariance tensor
```

Comment:
@uri-granta should we use check shapes now?

Reply:
Almost, but probably not quite yet. It's recently been moved out of GPflow into its own library, so I'll add a dependency on that while I upgrade trieste to use the latest GPflow (which is currently waiting on a GPflux release).